`ruff server` now supports `source.fixAll` source action

snowsignal opened the pull request:
## Summary

`ruff server` now has source action `source.fixAll` as an available code action.

This also fixes https://github.com/astral-sh/ruff/issues/10593 in the process of revising the code for quick fix code actions.

## Test Plan



https://github.com/astral-sh/ruff/assets/19577865/f4c07425-e68a-445f-a4ed-949c9197a6be

MichaReiser commented:
What do you think of implementing each checkbox in each own PR instead of having one large PR that implements all the features at once? To me, the individual features seem useful on its own and smaller PRs are easier to review, and tend to have a shorter lead time.

charliermarsh commented:
👍 Smaller PRs means we can merge faster and even release these out in the wild faster too.

MichaReiser commented on crates/ruff_server/src/lint.rs:
```diff
@@ -95,6 +92,7 @@ fn to_lsp_diagnostic(
                 serde_json::to_value(&DiagnosticFix {
                     kind: kind.clone(),
                     fix,
+                    code: format!("{}", rule.noqa_code()),
```
Nit
```suggestion
                    code: rule.noqa_code().to_string(),
```

MichaReiser commented on crates/ruff_server/src/server.rs:
```diff
@@ -191,6 +191,7 @@ impl Server {
                     code_action_kinds: Some(vec![
                         CodeActionKind::QUICKFIX,
                         CodeActionKind::SOURCE_ORGANIZE_IMPORTS,
+                        CodeActionKind::SOURCE_FIX_ALL,
```
Ruff's current LSP also supports a ruff specific `fixAll` action. It's [documented here](https://github.com/astral-sh/ruff-vscode?tab=readme-ov-file#configuring-vs-code) towards the end.

```json
{
  "[python]": {
    "editor.codeActionsOnSave": {
      "source.fixAll.ruff": "explicit",
      "source.organizeImports.ruff": "explicit"
    }
  }
}
```

MichaReiser commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,222 @@
+fn available_code_actions(
+    action_filter: Option<Vec<CodeActionKind>>,
+) -> BTreeSet<SupportedCodeAction> {
+    const DEFAULT_ACTIONS: &[SupportedCodeAction] = &[
+        SupportedCodeAction::QuickFix,
+        SupportedCodeAction::FixAll,
+        // SupportedCodeAction::OrganizeImports
+    ];
```
Nit: Should we move this into a `all()` method on `SupportCodeAction`. It could be easier to spot (and reuse) in case new variants are added.

MichaReiser commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,222 @@
+/// the list is filtered.
+fn available_code_actions(
+    action_filter: Option<Vec<CodeActionKind>>,
+) -> BTreeSet<SupportedCodeAction> {
```
Nit: We could consider using a bitflag for this, considering that the action kinds are a fixed set. But I'm also okay keeping with we have, avoiding the allocation here is probably overkill.

MichaReiser commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,222 @@
+        .then(move || {
+            edits_made
+                .into_iter()
+                .flat_map(|edit| edit.document_edits.iter())
+        })
+        .map(|changes| {
+            vec![
+                types::CodeActionOrCommand::CodeAction(types::CodeAction {
+                    title: format!("{DIAGNOSTIC_NAME}: Fix all auto-fixable problems"),
+                    diagnostics: Some(diagnostics_fixed),
+                    kind: Some(types::CodeActionKind::SOURCE_FIX_ALL),
                     edit: Some(types::WorkspaceEdit {
-                        document_changes: Some(types::DocumentChanges::Edits(changes)),
+                        document_changes: Some(types::DocumentChanges::Edits(
+                            changes.cloned().collect(),
+                        )),
                         ..Default::default()
                     }),
                     ..Default::default()
-                }))
-            })
-            .collect();
+                }), // TODO: implement command handler for the server
+                    /*
+                    types::CodeActionOrCommand::Command(types::Command {
+                        ...
+                    }
+                     */
+            ]
+        })
+        .into_iter()
+        .flatten()
+}
```
Nice trick to have a single iterator type. 

Nit: I do struggle reading this code. A lot is going on, including flattening. I wonder if the complexity is worth it to avoid a single `Vec` allocation, considering that we end up allocating a vec on line 174 anyway (we could avoid that by using an array instead).

I do find the non iterator version easier to reason about (I believe that's the equivalent code but I find it hard to say because I don't fully understand the iterator version)

````
    let edits_made: Vec<_> = edits
        .filter(|edit| {
            edit.diagnostic_fix
                .fix
                .applies(ruff_diagnostics::Applicability::Safe)
        })
        .collect();

    if edits_made.is_empty() {
        return Vec::new();
    }

    let diagnostics_fixed = edits_made
        .iter()
        .map(|edit| edit.original_diagnostic.clone())
        .collect();

    vec![
        types::CodeActionOrCommand::CodeAction(types::CodeAction {
            title: format!("{DIAGNOSTIC_NAME}: Fix all auto-fixable problems"),
            diagnostics: Some(diagnostics_fixed),
            kind: Some(types::CodeActionKind::SOURCE_FIX_ALL),
            edit: Some(types::WorkspaceEdit {
                document_changes: Some(types::DocumentChanges::Edits(
                    edits_made
                        .into_iter()
                        .flat_map(|edit| edit.document_edits.iter())
                        .cloned()
                        .collect(),
                )),
                ..Default::default()
            }),
            ..Default::default()
        }), // TODO: implement command handler for the server
            /*
            types::CodeActionOrCommand::Command(types::Command {
                ...
            }
             */
    ]
```

Edit: I think for now it would even be sufficient to return a single `Option`.

MichaReiser commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,222 @@
+    url: &'d Url,
+    encoding: PositionEncoding,
+    version: DocumentVersion,
+    diagnostics: impl Iterator<Item = types::Diagnostic> + 'd,
```
Nit: I think I would just go with a `vec` here to keep things simple. We don't have any usages today where we want to pass an iterator
```suggestion
    diagnostics: Vec<types::Diagnostic>,
```

MichaReiser commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,222 @@
+) -> impl Iterator<Item = CodeActionOrCommand> + 'd {
+    edits.map(|edit| {
+        let code = &edit.diagnostic_fix.code;
+        let title = edit
+            .diagnostic_fix
+            .kind
+            .suggestion
+            .clone()
+            .unwrap_or(edit.diagnostic_fix.kind.name.clone());
```
Nit: You can avoid cloning here by using `deref`
```suggestion
        let title = edit
            .diagnostic_fix
            .kind
            .suggestion
            .as_deref()
            .unwrap_or(&edit.diagnostic_fix.kind.name);
```

MichaReiser commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,222 @@
+
+        let mut response: types::CodeActionResponse = types::CodeActionResponse::default();
+
+        for action in available_actions {
+            match action {
+                SupportedCodeAction::QuickFix => response.extend(quick_fix(edits.iter())),
+                SupportedCodeAction::FixAll => response.extend(fix_all(edits.iter())),
+                SupportedCodeAction::OrganizeImports => {
+                    todo!("Implement the `source.organizeImports` code action")
+                }
+            }
+        }
```
Nit: Considering that `available_actions` is a set (or bitset, see below), I would probably rewrite it as

```
			  // Add a quick fix for the diagnostic
        if available_actions.contains(&SupportedCodeAction::QuickFix) {
            response.extend(quick_fix(edits.iter()));
        }

				// Add a single fix all action for all diagnostics
        if available_actions.contains(&SupportedCodeAction::FixAll) {
            response.extend(fix_all(edits.iter()));
        }

        if available_actions.contains(&SupportedCodeAction::OrganizeImports) {
            todo!("Implement the `source.organizeImports` code action")
        }
```

I find this more readable because it makes the ordering explicit in code (no need to implement `Ord`) and allows for inline comments.

MichaReiser commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,222 @@
-                    kind: Some(types::CodeActionKind::QUICKFIX),
+        // compute the associated document edits for each diagnostic
+        // these will get re-used when building the actual code actions afterwards
+        let edits = diagnostic_edits(
```
I'm struggling a bit with the terminology: 

* A diagnostic has a list of edits
* The diagnostic gets converted into a `DiagnosticEdit` but it itself has a list of edits
* `diagnostic_edits` returns a vec of `DiagnosticEdit` (which is a list of edits). 

My recommendation would have been to rename `DiagnosticEdit` to `DiagnosticFix` but that type already exists and that type also stores a list of edits (but different edits?). 

I don't have a good recommendation on how to simplify this. Maybe rename `DiagnosticFix` to `SerializedFix` or `SerializedDiagnosticFix` to make it clear it's what is stored on `Diagnostic` or `SerializedDiagnosticData`. I mainly want to point out that I'm overwhelmed by this and struggle to understand the different data structures.

snowsignal replied:
I've made renamings similar to what you suggested, with one exception: I changed `DiagnosticFix` to `AssociatedDiagnosticData` because I think that describes what its actual role is as a serialized structure. `SerializedFix` may imply that it contains serialized values, which it doesn't.

I also reworked `DiagnosticEdit` (now `DiagnosticFix`) to only include the relevant fields from `AssociatedDiagnosticData` so it's a bit less confusing.

MichaReiser commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,222 @@
+        .filter_map(Result::transpose)
+}
+
+fn quick_fix<'d>(
+    edits: impl Iterator<Item = &'d DiagnosticEdit> + 'd,
+) -> impl Iterator<Item = CodeActionOrCommand> + 'd {
```
Nit: I would remove the generic type considering that `edits` is always a slice
```suggestion
fn quick_fix<'d>(
    edits: &['d DiagnosticEdit],
) -> impl Iterator<Item = CodeActionOrCommand> + 'd {
```

It simplifies the signature a great deal

MichaReiser reviewed (approved):
Nice, now we have almost feature parity with the old extension for Pyton files. 

This looks good to me. I have a few small nits. Feel free to disregard them; this is mainly a matter of personal preference. 

I do think there's an opportunity to improve the naming. It took me a while to understand all the different types involved. I made a suggestion inline. 

Once you add the test plan to the PR (which you planned on doing anyway), this is good to go.

snowsignal commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,222 @@
+) -> impl Iterator<Item = crate::server::Result<DiagnosticEdit>> + 'd {
+    diagnostics
+        .map(move |diagnostic| {
+            let Some(data) = diagnostic.data.clone() else {
```
My only concern is that this would make the list of 'fixed' diagnostics that we return no longer perfectly match the original diagnostics the client has, since they'll be missing the `data` field. Unfortunately the specification isn't clear whether it has to be an exact match or whether it matches on certain fields. Since this is just supposed to be a hint though, it's probably fine.

MichaReiser replied:
Yeah. That's why I'm unsure if it's okay to do so. I'm very surprised that the response must contain the entire diagnostic and not just an ID.

snowsignal commented:
I've addressed all the suggestions and rebased against `main` to get the fix for https://github.com/astral-sh/ruff/issues/10618. All that's left is the test plan video.

dhruvmanila commented on crates/ruff_server/src/server.rs:
```diff
@@ -196,6 +196,7 @@ impl Server {
                     code_action_kinds: Some(vec![
                         CodeActionKind::QUICKFIX,
                         CodeActionKind::SOURCE_ORGANIZE_IMPORTS,
+                        CodeActionKind::SOURCE_FIX_ALL,
```
We will need to advertise to the client that we support Ruff scoped code actions during initialization. Refer https://github.com/astral-sh/ruff-lsp/blob/187d7790be0783b9ac41ce025a724cf389bf575c/ruff_lsp/server.py#L852-L869

MichaReiser commented:
Thanks for addressing all the feedback.

dhruvmanila commented on crates/ruff_server/src/server/api/requests/code_action_resolve.rs:
```diff
@@ -0,0 +1,103 @@
+) -> crate::Result<types::CodeAction> {
+    let diagnostics = crate::lint::check(document, linter_settings, encoding);
+
+    let fixes = crate::lint::fixes_for_diagnostics(
+        document,
+        url,
+        encoding,
+        document.version(),
+        diagnostics,
+    )
+    .collect::<crate::Result<Vec<_>>>()?;
```
Do you think we should just return an owned vector of fixes here instead of an iterator impl? Returning an impl Iterator makes me wonder if there's any potential use case, you will have more knowledge on this.

snowsignal replied:
Returning an `impl Iterator` tends to be my default when writing functions that use iterators (as opposed to collecting them preemptively) because it gives the caller more control over the allocation and execution of the iterator. I'm not opposed to `.collect()` in a function body if it makes the code easier to read though, so I'm fine with changing the signature.

dhruvmanila reviewed (commented):
This is great work!

I just have one recommendation which is to document the types and functions. This would be really helpful. It's does not need to be done in this PR itself.

I'm going to test this out locally in Neovim now and approve then.

dhruvmanila reviewed (approved):
I just tested it in Neovim. It works really well.

I have noticed one minor thing but it could very well be something else. When I run the fixAll code action and then undo the edit via the editor, the cursor jumps to the first line of the file. This doesn't happen when I do so through `ruff-lsp`. This is a bit of inconvenience but I don't think it's related to your change. I can look into why this is happening. It could very well be an editor bug.

Here's a video showcasing that: the first code action applied is via `ruff-lsp` while the second one is the release build of this branch.

https://github.com/astral-sh/ruff/assets/67177269/c343e780-3d29-49c2-aabf-a17f7c1ecbb9

MichaReiser commented on crates/ruff_server/src/session/settings.rs:
```diff
@@ -0,0 +1,32 @@
+use lsp_types::ClientCapabilities;
+
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub(crate) struct ServerSettings {
+    pub(crate) capabilities: ExtensionCapabilities,
+}
```
I share the sentiment here, but I think it's mainly a naming issue than anything else:

* It's unexpected to find server settings in the session
* The settings (as they are today) configure only the client capabilities

I think my recommendation for today would be to

a) Remove `ServerSettings` and directly use `ClientCapabilities` or `ResolvedClientCapabilities` (inline today's `ExtensionCapabilities`). 
b) Rename to `SessionSettings`?

MichaReiser commented on crates/ruff_server/src/server.rs:
```diff
@@ -73,10 +73,10 @@ impl Server {
             threads,
             worker_threads,
-            session: Session::new(&server_capabilities, &workspaces)?,
+            session: Session::new(&client_capabilities, &server_capabilities, &workspaces)?,
+            client_capabilities,
```
Would it be sufficient to only store the client capabilities on the session, or why do we need to store it on the sessoin and server?

snowsignal replied:
We need a separate reference to `client_capabilities` to pass into `Server::try_register_capabilities` because the session is already borrowed mutably by the scheduler.

MichaReiser commented on crates/ruff_server/src/session/settings.rs:
```diff
@@ -0,0 +1,32 @@
+            code_action_deferred_edit_resolution: client_capabilities
+                .text_document
+                .as_ref()
+                .and_then(|doc_settings| doc_settings.code_action.as_ref())
+                .and_then(|code_action_settings| code_action_settings.resolve_support.as_ref())
+                .is_some_and(|resolve_support| resolve_support.properties.contains(&"edit".into())),
+        }
```
We also need to check that the extension has data support:

> This is also guarded by an additional client capability codeAction.dataSupport. In general, a client should offer data support if it offers resolve support. It should also be noted that servers shouldn’t alter existing attributes of a code action in a codeAction/resolve request.

MichaReiser reviewed (changes requested):
I've been a bit surprised by the amount of changes in your last commit.  Were the changes necessary to address a bug in the implementation? 

If not, then I would recommend for the future to merge the existing PR (it's functional) and create a new PR with the changes instead. This has a few advantages in my view:

* It gives you place to explain and motivate your change. This prevents the situation I'm in now where it's unclear why we implemented a new resolve handler. What is it for, what does it enable? Why the added complexity? 
* It avoids that reviewers have to re-review the entire PR a second time. Instead, they can focus on the latest changes (in context of what you summarized)


If you make such fundamental changes to existing PRs (because it's required to address a bug), then I recommend you to explicitly re-request review. It was unclear to me why the PR wasn't merged this morning and I only understood why after seeing @dhurvmanila's comments

Would you mind explaining the most recent changes in the PR summary. How does `resolve` and code action work together. How do we handle clients that don't support these client capabilities. I'm sure I could read this out from the code but that adds an extra burden to the reviewer because:

* I have to figure out the actual behavior from code
* I have to figure out the expected behavior my self (or the behavior you intended)


That makes it more difficult for me to make a useful review.

snowsignal commented:
@MichaReiser My apologies for not explaining the reasons I made these changes more clearly. I'll try to go over the issue with this PR and the need to implement `codeAction/resolve` as a part of that:

1. When we generate a list of code actions, we're given a list of diagnostics to generate those code actions for. My first pass implementation made the assumption that this diagnostic list would be for the entire file, and so the fixes used to create the `fixAll` workspace edit were taken directly from the diagnostics provided.
2. However, this isn't how the code action endpoint works, most of the time. The client usually asks for code actions based on diagnostics in a specific range, which for VS Code is usually just the underlined diagnostic the cursor is pointing to. 
3. The way the existing LSP (`ruff-lsp`) solves this issue is by always making `source.fixAll` (and `source.organizeImports`) available as a code action, and attempting to defer calculating the `edit` field until the source action is actually being called.
4. That's essentially what I changed this PR to do - we check if the client supports deferred code action resolution, and defer calculating the diagnostic for the entire file. If the client supports this deferred resolution, we set the `edit` field to `None` when building the code action as part of the `textDocument/codeAction` request. It's then the client's job to call `codeAction/resolve` once the code action is about to be used, where we populate the `edit` field with the edits to make, if safe fixes exist for the file. If the client doesn't support deferred resolution, we populate the `edit` field during the `textDocument/codeAction` request.

I hope that clears up the confusion around these changes. I'll try to have these explanations prepared in advance for the future.

snowsignal commented on crates/ruff_server/src/lint.rs:
```diff
@@ -78,6 +91,56 @@ pub(crate) fn check(
         .collect()
 }
 
+pub(crate) fn fixes_for_diagnostics<'d>(
+    document: &'d crate::edit::Document,
+    url: &'d lsp_types::Url,
+    encoding: PositionEncoding,
+    version: crate::edit::DocumentVersion,
+    diagnostics: Vec<lsp_types::Diagnostic>,
+) -> impl Iterator<Item = crate::Result<DiagnosticFix>> + 'd {
```
I generally like to keep `use` statements to a minimum, but I can agree it's a bit verbose. I'll try to remove fully qualified paths where the type in question has an obvious origin.

snowsignal commented on crates/ruff_server/src/session/settings.rs:
```diff
@@ -0,0 +1,32 @@
+    pub(crate) capabilities: ExtensionCapabilities,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub(crate) struct ExtensionCapabilities {
+    pub(crate) code_action_deferred_edit_resolution: bool,
+}
```
My intent for `ExtensionSettings` and its sub-structure `ExtensionCapabilities` was to reflect the configuration being set on the client side (AKA the 'editor extension'), along with the capabilities of the client. That being said, `ResolvedClientCapabilities` is probably a better name for this!

snowsignal commented on crates/ruff_server/src/server.rs:
```diff
@@ -236,3 +237,69 @@ impl Server {
+        }
+    }
+
+    fn makes_available(self) -> AvailableCodeActions {
```
We take `self` by value here because `Self` is `Copy`-able and clippy complains if you take it by reference. It's not meant to be a conversion method, but maybe that needs to be clarified.

snowsignal commented on crates/ruff_server/src/server/api/requests/code_action_resolve.rs:
```diff
@@ -0,0 +1,103 @@
+    ) -> Result<types::CodeAction> {
+        let document = snapshot.document();
+
+        let supported: SupportedCodeActionKind = action
+            .kind
+            .clone()
+            .ok_or(anyhow::anyhow!("No kind was given for code action"))
+            .with_failure_code(ErrorCode::InvalidParams)?
+            .try_into()
+            .map_err(|()| anyhow::anyhow!("Code action was of an invalid kind"))
+            .with_failure_code(ErrorCode::InvalidParams)?;
+        let available_action = supported.makes_available();
+
+        // ensures that only one code action kind was made available
+        debug_assert!(
+            (available_action & (available_action - AvailableCodeActions::all()))
+                == AvailableCodeActions::empty()
+        );
+
+        if available_action == AvailableCodeActions::SOURCE_FIX_ALL {
+            resolve_edit_for_fix_all(
+                action,
```
I wanted to avoid duplicating the logic of `makes_available`, since I was concerned about a misalignment in logic if we added new variants of code action kinds to be accounted for. However, I'd be fine with using a match statement like this if we make sure it's exhaustive, so that adding new variants would force us to update the `match` logic.

snowsignal commented on crates/ruff_server/src/server/api/requests/code_action_resolve.rs:
```diff
@@ -0,0 +1,103 @@
+    ) -> Result<types::CodeAction> {
+        let document = snapshot.document();
+
+        let supported: SupportedCodeActionKind = action
+            .kind
+            .clone()
+            .ok_or(anyhow::anyhow!("No kind was given for code action"))
+            .with_failure_code(ErrorCode::InvalidParams)?
+            .try_into()
+            .map_err(|()| anyhow::anyhow!("Code action was of an invalid kind"))
+            .with_failure_code(ErrorCode::InvalidParams)?;
+        let available_action = supported.makes_available();
+
+        // ensures that only one code action kind was made available
+        debug_assert!(
+            (available_action & (available_action - AvailableCodeActions::all()))
+                == AvailableCodeActions::empty()
+        );
+
+        if available_action == AvailableCodeActions::SOURCE_FIX_ALL {
+            resolve_edit_for_fix_all(
+                action,
+                document,
+                snapshot.url(),
+                &snapshot.configuration().linter,
+                snapshot.encoding(),
+            )
+            .with_failure_code(ErrorCode::InternalError)
```
I'm not sure what this suggestion is for - it just seems to be removing most of the `resolve_edit_for_fix_all` call, which wouldn't compile.

MichaReiser commented:
Thanks for the explanation. That's helfpul context for reviewing the PR

> If the client doesn't support deferred resolution, we populate the edit field during the textDocument/codeAction request.

Does that mean that `fix all` for clients not supporting `resolve` doesn't fix all diagnostics but only the diagnostics in view (the trimmed-down diagnostics selected by the editor?)

MichaReiser commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,129 @@
+    let mut action = types::CodeAction {
+        title: format!("{DIAGNOSTIC_NAME}: Fix all auto-fixable problems"),
+        kind: Some(types::CodeActionKind::SOURCE_FIX_ALL),
+        // This will be resolved later
+        edit: None,
```
Nit: 
```suggestion
        // The editor will request the edit in a `CodeActionsResolve` request
        edit: None,
```

snowsignal commented:
> Does that mean that fix all for clients not supporting resolve doesn't fix all diagnostics but only the diagnostics in view (the trimmed-down diagnostics selected by the editor?)

I should have been more clear on that: we _do_ fix all diagnostics, it's just less performant because we have to lint the file right then instead of deferring that to when it's actually used.

MichaReiser commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,129 @@
-        let url = snapshot.url();
-        let encoding = snapshot.encoding();
-        let version = document.version();
-        let actions: Result<Vec<_>> = params
-            .context
-            .diagnostics
-            .into_iter()
-            .map(|diagnostic| {
-                let Some(data) = diagnostic.data else {
-                    return Ok(None);
-                };
-                let diagnostic_fix: crate::lint::DiagnosticFix = serde_json::from_value(data)
-                    .map_err(|err| anyhow::anyhow!("failed to deserialize diagnostic data: {err}"))
-                    .with_failure_code(lsp_server::ErrorCode::ParseError)?;
-                let edits = diagnostic_fix
-                    .fix
-                    .edits()
+
+        // We need to resolve the `edit` field now if we can't defer resolution to later
+        action = resolve_edit_for_fix_all(
+            action,
+            document,
+            snapshot.url(),
+            &snapshot.configuration().linter,
+            snapshot.encoding(),
+        )?;
+    }
+
+    Ok(types::CodeActionOrCommand::CodeAction(action))
```
It's a bit confusing that we initialize `action` only to then override it. To me it seems the only motivation for doing this is so that the `action` can be passed to `resolve_edit_for_fix_all`, which sets `action.edit`. I think I would restructure this to
```suggestion
    let (edit, data) = if snapshot.resolved_client_capabilities().code_action_deferred_edit_resolution {
        // This will be resolved later
        (None, Some(serde_json::to_value(snapshot.url()).expect("document url to serialize")))
    } else {
        (resolve_edit_for_fix_all(document, snapshot.url(), &snapshot.configuration.linter, snapshot.encoding()), None)
    };

    let action = types::CodeAction {
        title: format!("{DIAGNOSTIC_NAME}: Fix all auto-fixable problems"),
        kind: Some(types::CodeActionKind::SOURCE_FIX_ALL),
        edit: None,
        data,
        ..Default::default()
    };

    Ok(types::CodeActionOrCommand::CodeAction(action))
```

This way it's clearer what's different between the two paths and it avoids setting `data` for clients that don't support it.

MichaReiser commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,129 @@
+use super::code_action_resolve::resolve_edit_for_fix_all;
 
-impl super::RequestHandler for CodeAction {
+pub(crate) struct CodeActions;
```
Nit: It might be helpful to document the different code actions. What are they supposed to do and how do they work with other code actions.

dhruvmanila commented on crates/ruff_server/src/server/api/requests/code_action.rs:
```diff
@@ -1,81 +1,129 @@
+
+        let supported_code_actions = supported_code_actions(params.context.only);
+
+        tracing::error!("{supported_code_actions:?}");
```
Did this get accidentally checked-in or is it suppose to be at a different log level?

snowsignal replied:
This was checked in by accident, my bad.

MichaReiser commented:
Whoops. I directly commented on the commit https://github.com/astral-sh/ruff/commit/de6870df6528455b59adbc10cfe8247cac449e78 because that was the easiest to review the changes but it seems GitHub doesn't show these comments :( Anyway. Looks good to me.

dhruvmanila commented:
> Whoops. I directly commented on the commit [de6870d](https://github.com/astral-sh/ruff/commit/de6870df6528455b59adbc10cfe8247cac449e78) because that was the easiest to review the changes but it seems GitHub doesn't show these comments :( Anyway. Looks good to me.

I think that only works if you open the commit from the PR's "Files Changed" tab. There's a commit filter on the top of the diff view. Maybe that should render the comments correctly.
